metadata/src/utils/retain.rs

```diff
@@ -4,49 +4,278 @@
+                    for variant in variants {
+                        for field in &variant.fields {
+                            if self.seen_ids.insert(field.ty.id) {
+                                let ty = resolve_typ(metadata, field.ty.id);
+                                work_set.push_back(ty);
+                            }
+                        }
+                    }
+                }
+                TypeDef::Array(TypeDefArray { len: _, type_param })
+                | TypeDef::Sequence(TypeDefSequence { type_param })
+                | TypeDef::Compact(TypeDefCompact { type_param }) => {
+                    if self.seen_ids.insert(type_param.id) {
+                        let ty = resolve_typ(metadata, type_param.id);
+                        work_set.push_back(ty);
+                    }
+                }
+                TypeDef::Tuple(TypeDefTuple { fields }) => {
+                    for field in fields {
+                        if self.seen_ids.insert(field.id) {
+                            let ty = resolve_typ(metadata, field.id);
+                            work_set.push_back(ty);
+                        }
+                    }
+                }
+                TypeDef::Primitive(_) => (),
+                TypeDef::BitSequence(TypeDefBitSequence {
+                    bit_store_type,
+                    bit_order_type,
+                }) => {
+                    for typ in [bit_order_type, bit_store_type] {
+                        if self.seen_ids.insert(typ.id) {
+                            let ty = resolve_typ(metadata, typ.id);
+                            work_set.push_back(ty);
+                        }
+                    }
                 }
             }
         }
     }
 
-    if let Some(ty) = pallet.call_ty {
-        type_ids.insert(ty);
+    fn insert_and_collect_types(&mut self, metadata: &Metadata, ty: &PortableType) {
+        if self.seen_ids.insert(ty.id) {
+            self.collect_types(metadata, ty);
+        }
+    }
+
+    fn collect_extrinsic_types(&mut self, extrinsic: &ExtrinsicMetadata) {
+        let mut ids = Vec::from([
+            extrinsic.address_ty,
+            extrinsic.call_ty,
+            extrinsic.signature_ty,
+            extrinsic.extra_ty,
+        ]);
+
+        for signed in &extrinsic.signed_extensions {
+            ids.push(signed.extra_ty);
+            ids.push(signed.additional_ty);
+        }
+        for id in ids {
+            self.seen_ids.insert(id);
+        }
     }
 
-    if let Some(ty) = pallet.event_ty {
-        type_ids.insert(ty);
+    /// Collect all type IDs needed to represent the runtime APIs.
+    fn collect_runtime_api_types(&mut self, api: &RuntimeApiMetadataInner) {
+        for method in api.methods.values() {
+            for input in &method.inputs {
+                self.seen_ids.insert(input.ty);
+            }
+            self.seen_ids.insert(method.output_ty);
+        }
+    }
```

Comment:
These two functions don't recurse through any sub types. While I don't think anything should, if we did decide we wanted to recurse here then I guess these should, too

Reply:
pushed a change that does traverse, a few more types added to retained set on average.